### PR TITLE
sql: enable adding columns which are not written to the current primary index

### LINKED ISSUE #59149: sql: enable adding columns which are not written to the current primary index
**Is your feature request related to a problem? Please describe.**

This issue is a breakout of point 2 in https://github.com/cockroachdb/cockroach/issues/47989#issuecomment-760626388. Column backfills for changing the set of columns in a table have a number of issues outlined in https://github.com/cockroachdb/cockroach/issues/47989. The alternative is to build a new primary index and swap to it. We support this in theory with primary key changes, however, there's some slight nuance in that primary key changes require the set of columns in the new and old primary index are the same. This issue is concerned with the requirement that concurrent writer not write any new columns being backfilled to this new primary index to the old primary index.

The root of the problem is that primary index descriptors are handled specially. For secondary indexes, the columns stored in the value are primary index columns specified in `ExtraColumnIDs` for unique indexes and the columns in `StoreColumnIDs`. For primary indexes, however, all of the table columns are stored.

The hazard is that, if we don't change anything, the column in DELETE_AND_WRITE_ONLY will end up being written to the existing primary index.

**Describe the solution you'd like**

The thrust of the solution is twofold:

1) Update the table descriptor to encode the need to not write these columns to the current primary index.
2) Adopt this change where necessary.

#### Updated descriptor structure

I have two proposals here, one that I prefer but is riskier and one that is less invasive but worse. The latter does not in any 

* A)  Make primary indexes look like secondary indexes and populate their `StoreColumnIDs`.
    * *Pros*
        * This is nice for a variety of reasons. It's not obvious to me why we maintain this distinction save for legacy reasons. It should simplify index handling code. 
    * *Cons*
        * It's slightly less compact.
        * It'd be a migration that would affect a lot of code.

* B) Add another field to the table descriptor to encode the set of columns which should not be written to the primary index
   * *Pros*
      * Much less invasive
   * *Cons*
      * More cruft

#### Adopting change to descriptor structure.

Let's assume we're going to go with B) as it seems more tractable. One thing which will need to change is the code which writes rows. My current reading is that we might be able to isolate this change to legitimately just [this function](https://github.com/cockroachdb/cockroach/blob/3e2a4e7f8891d153b4785a6a337522481c0ffedb/pkg/sql/row/helper.go#L133-L156). @yuzefovich could I ask you to verify that claim and generally have a look at this issue?

The bigger unknown for me is what changes would need to be made in the optimizer. @RaduBerinde could I ask you to give review this and provide some guidance. 

**Additional context**

Something must be done here to realize https://github.com/cockroachdb/cockroach/issues/47989. More pressingly, we are working to not include support for column backfills in the new schema change rewrite so realistically something needs to be done here in the next few weeks.

### PR DESCRIPTION
**Pull request by postamar**
row,schemachanger: use StoreColumnIDs/Names in primary index
    
    In a recent commit, the StoreColumnIDs and StoreColumnNames slices in
    primary indexes were populated when previously they had simply been
    empty. We simply assumed that all non-virtual columns in a table would
    be stored in the primary index: primary key columns in the key, the rest
    in the value.
    
    This commit breaks that assumption by using the StoreColumnIDs slice to
    determine what goes into the primary index. This makes it possible for
    the new schema changer to add columns safely, preventing unwanted writes
    to the old primary index while the schema change is underway.
    
    Fixes #59149.
    
    Release note: None

    sql,tabledesc: add new IndexDescriptorVersion for primary indexes
    
    Previously, the IndexDescriptorVersion type was only used to describe
    the encoding of secondary indexes. This commit adds a new value for
    use in primary indexes, PrimaryIndexWithStoredColumnsVersion, to signify
    that the StoredColumnIDs and StoredColumnNames slices are populated
    correctly.
    
    Previously, these slices did not need to be populated at all. This is
    because the set of columns comprising the primary index of a table is
    assumed to be all non-virtual columns of that table. Our upcoming work on
    the new schema changer will require us to violate that assumption
    however. This commit is in preparation of that change.
    
    In our effort to make meaningful the concept of stored columns in
    primary indexes, this commit also changes the contents of the
    information_schema.statistics table. As a result, SHOW INDEXES and SHOW
    COLUMNS behave the same way regardless of whether an index is primary or
    secondary.
    
    Release note (sql change): The contents of the statistics table in the
    information schema have changed, therefore so have the results of SHOW
    INDEX and SHOW COLUMNS. A column which is not in the primary key will
    now be listed as belonging to the primary index as a stored column.
    Previously, it was simply not listed as belonging to the primary index.

### CODE REVIEW

**Comment by postamar** [comment]
Rebased to fix merge conflicts.

**Review by ajwerner — APPROVED** [review]
:lgtm: just small things. Also looks like you've still got some merge skew. 

> Reviewed 65 of 68 files at r1, 4 of 12 files at r2, 14 of 14 files at r3, 12 of 12 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66599#-:-McivbnW0esDBIowxoC3:b-7iy4do)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @postamar)

___
*[pkg/sql/truncate.go, line 226 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66599#-McikVT32_DrXbpvmXn9:-McikVT32_DrXbpvmXnA:bb1z2xt) ([raw file](https://github.com/cockroachdb/cockroach/blob/3efb8850d15e3c008b2be2fe0d55c42dcc196088/pkg/sql/truncate.go#L226)):*
<details><summary><i alt="summary">Quoted 8 lines of code…</i></summary>

> ```
> 	// Exit early with an error if the table is undergoing a new-style schema
> 	// change, before we try to get job IDs and update job statuses later. See
> 	// createOrUpdateSchemaChangeJob.
> 	if tableDesc.NewSchemaChangeJobID != 0 {
> 		return pgerror.Newf(pgcode.ObjectNotInPrerequisiteState,
> 			"cannot perform a schema change on table %q while it is undergoing a new-style schema change",
> 			tableDesc.GetName(),
> 		)
> ```
</details>

Should this get pushed into `checkTableForDisallowedMutationsWithTruncate`?
___
*[pkg/sql/catalog/table_elements.go, line 168 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66599#-Mcisb1f7XVVAk9h5zFU:-Mcisb1f7XVVAk9h5zFV:b-t2ffgh) ([raw file](https://github.com/cockroachdb/cockroach/blob/3efb8850d15e3c008b2be2fe0d55c42dcc196088/pkg/sql/catalog/table_elements.go#L168)):*
> ```Go
> 	GetStoredColumnID(storedColumnOrdinal int) descpb.ColumnID
> 	GetStoredColumnName(storedColumnOrdinal int) string
> 	HasOldStoredColumns() bool
> ```

I wish this concept (which I understand is not new) was better commented and explained somewhere. There's a small comment in `tabledesc` but it's not satisfying. 
___
*[pkg/sql/catalog/tabledesc/structured.go, line 680 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66599#-McisohE7XwHT9GHqguk:-McisohE7XwHT9GHqgul:b-aw7emb) ([raw file](https://github.com/cockroachdb/cockroach/blob/3efb8850d15e3c008b2be2fe0d55c42dcc196088/pkg/sql/catalog/tabledesc/structured.go#L680)):*
> ```Go
> 
> 		indexHasOldStoredColumns := idx.HasOldStoredColumns()
> 		// Need to clear KeySuffixColumnIDs and StoreColumnIDs
> ```

Can you use more words to explain what's going on here? I know this existed before but this code is hard to follow. 
___
*[pkg/sql/catalog/tabledesc/table_desc.go, line 190 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66599#-McitLZI4qsl7eYLUZnV:-McitLZI4qsl7eYLUZnW:bctqgmj) ([raw file](https://github.com/cockroachdb/cockroach/blob/3efb8850d15e3c008b2be2fe0d55c42dcc196088/pkg/sql/catalog/tabledesc/table_desc.go#L190)):*
> ```Go
> 		return true
> 	}
> ```

Should this be asserting something about the format of `idx`?
___
*[pkg/sql/schemachanger/schemachanger_test.go, line 417 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/66599#-Mciw1BNDvjE0-mj0SO8:-Mciw1BNDvjE0-mj0SO9:b-sjji2l) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8ebd2bb4c59b0028911398f75eae7433c1cf7d1/pkg/sql/schemachanger/schemachanger_test.go#L417)):*
> ```Go
> func TestInsertDuringAddColumnNotWritingToCurrentPrimaryIndex(t *testing.T) {
> 	defer leaktest.AfterTest(t)()
> 	defer sqltestutils.SetTestJobsAdoptInterval()()
> ```

This no longer exists.

**Comment by postamar** [comment]
Thanks for the review. I pushed a fix to the merge skew concurrently to you posting your review. I'll address your comments ASAP.

**Review by postamar — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66599)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner)

___
*[pkg/sql/truncate.go, line 226 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66599#-McikVT32_DrXbpvmXn9:-McizbIL4IqasGN18CU2:b-i2tph8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3efb8850d15e3c008b2be2fe0d55c42dcc196088/pkg/sql/truncate.go#L226)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 	// Exit early with an error if the table is undergoing a new-style schema
> 	// change, before we try to get job IDs and update job statuses later. See
> 	// createOrUpdateSchemaChangeJob.
> 	if tableDesc.NewSchemaChangeJobID != 0 {
> 		return pgerror.Newf(pgcode.ObjectNotInPrerequisiteState,
> 			"cannot perform a schema change on table %q while it is undergoing a new-style schema change",
> 			tableDesc.GetName(),
> 		)
> ```
Should this get pushed into `checkTableForDisallowedMutationsWithTruncate`?
</blockquote></details>

I thought about this and my rationale for not putting it in there was that this test didn't really involve mutations. Also it's new-schemachanger stuff and that function is old-schemachanger stuff, I guessed that it would be better to keep them separate.

I don't feel strongly about any of this so if you prefer I'll push it there.
___
*[pkg/sql/catalog/table_elements.go, line 168 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66599#-Mcisb1f7XVVAk9h5zFU:-Mcj-MSaCEM28wDfUjA8:baw789p) ([raw file](https://github.com/cockroachdb/cockroach/blob/3efb8850d15e3c008b2be2fe0d55c42dcc196088/pkg/sql/catalog/table_elements.go#L168)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I wish this concept (which I understand is not new) was better commented and explained somewhere. There's a small comment in `tabledesc` but it's not satisfying. 
</blockquote></details>

I completely agree. I poked around and found this, which has a bit more substance to it: https://github.com/cockroachdb/cockroach/blob/eb8e1c8db70fe61356319aff97a6cded8720d9d2/pkg/sql/information_schema.go#L1117-L1128

I'll improve the comment on `HasOldStoredColumns` in light of this.
___
*[pkg/sql/catalog/tabledesc/structured.go, line 680 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66599#-McisohE7XwHT9GHqguk:-Mcj0JsB60QanvH24i_d:bv71ab0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3efb8850d15e3c008b2be2fe0d55c42dcc196088/pkg/sql/catalog/tabledesc/structured.go#L680)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Can you use more words to explain what's going on here? I know this existed before but this code is hard to follow. 
</blockquote></details>

Yes this is almost certainly an omission on my part, sorry.
___
*[pkg/sql/catalog/tabledesc/table_desc.go, line 190 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66599#-McitLZI4qsl7eYLUZnV:-Mcj0Mx_EpY7IaUMEAQr:bssvflb) ([raw file](https://github.com/cockroachdb/cockroach/blob/3efb8850d15e3c008b2be2fe0d55c42dcc196088/pkg/sql/catalog/tabledesc/table_desc.go#L190)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Should this be asserting something about the format of `idx`?
</blockquote></details>

At this point the index is primary and is therefore assumed to be in the upgraded version. Adding an assertion here would be redundant due to on-commit descriptor validation... aaaand I just realized that we don't actually have a validation check for that! I'll address this.
___
*[pkg/sql/schemachanger/schemachanger_test.go, line 417 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/66599#-Mciw1BNDvjE0-mj0SO8:-Mcj57GW80-tPYAWz8r8:b-ucfkn4) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8ebd2bb4c59b0028911398f75eae7433c1cf7d1/pkg/sql/schemachanger/schemachanger_test.go#L417)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This no longer exists. 
</blockquote></details>

Thanks.
